internal/policy/container/has_modified_files.go

```diff
@@ -45,118 +48,155 @@ func (p *HasModifiedFilesCheck) Validate(ctx context.Context, imgRef image.Image
-// This avoids establishing a baseline of files installed in the original layer with an empty list. An example
-// of when this should occur is in cases where the image was built FROM scratch, such as ubi-micro. This is required
-// to ensure we don't attempt to validate against an empty layer.
-func (p *HasModifiedFilesCheck) dropFirstLayerIfEmpty(files [][]string) ([][]string, bool) {
-	var dropped bool
-	if len(files[0]) == 0 {
-		files = files[1:] // shift the empty layer out.
-		dropped = true
+func extractPackageNameVersion(pkgList []*rpmdb.PackageInfo) map[string]struct{} {
+	pkgNameList := make(map[string]struct{}, len(pkgList))
+	for _, pkg := range pkgList {
+		pkgNameList[fmt.Sprintf("%s-%s", pkg.Name, pkg.Version)] = struct{}{}
 	}
-
-	return files, dropped
+	return pkgNameList
 }
 
 // validate compares the list of LayerFiles and PackageFiles to see what PackageFiles
 // have been modified within the additional layers.
 //
 //nolint:unparam // ctx is unused. Keep for future use.
 func (p *HasModifiedFilesCheck) validate(ctx context.Context, packageFilesRef *packageFilesRef) (bool, error) {
-	logger := logr.FromContextOrDiscard(ctx)
-
-	layerFiles := packageFilesRef.LayerFiles
-	packageFiles := packageFilesRef.PackageFiles
-
-	// Determine the list of files that were a part of the base layer.
-	baseLayer := make(map[string]struct{}, len(layerFiles[0]))
-	for _, filename := range layerFiles[0] {
-		if _, ok := packageFiles[filename]; ok {
-			baseLayer[filename] = struct{}{}
+	modifiedFilesDetected := false
+	layerFileList := packageFilesRef.LayerFiles
+	layerPackages := packageFilesRef.LayerPackages
+	layerPackageFiles := packageFilesRef.LayerPackageFiles
+	layerIDs := packageFilesRef.Layers
+
+	lastLayerPackageFiles := make(map[string]string, 0)
+	for _, layerID := range layerIDs {
+		if len(layerPackages[layerID]) == 0 {
+			lastLayerPackageFiles = nil
+			continue
 		}
-	}
-	layers := layerFiles[1:]
 
-	// Look for modifications in subsequent layers by determining
-	// if a file in a base layer exists in a subsequent layer.
-	modifiedFilesDetected := false
-	for _, layer := range layers {
-		for _, file := range layer {
-			if _, ok := baseLayer[file]; ok {
-				// This means the files exists in the base layer. This is a fail.
-				logger.V(log.DBG).Info("modified file detected", "filename", file)
+		for _, layerFile := range layerFileList[layerID] {
+			lastPackageVer, found := lastLayerPackageFiles[layerFile]
+			if !found {
+				// Either the file was not in the last layer (new package, which is allowed),
+				// or is not owned by a package, so ignore
+				continue
+			}
+			currentPackageVer, found := layerPackageFiles[layerID][layerFile]
+			if !found {
+				// Package was removed? This is bad too. Set modified files, and exit
 				modifiedFilesDetected = true
+				break
 			}
+			if lastPackageVer == currentPackageVer {
+				// This means the file was changed, but the package was not. Stinky.
```

Comment:
Stinky!

Reply:
I liked that too, and the other comments in this block. It covers the things we've talked about nicely IMO.